TST: spatial.transform: Add array API standard support for testing

amacati opened the pull request:
Add Array API support to `scipy.spatial.transform` tests in preparation of #22777 

#### Reference issue
#22500, #22777 

#### What does this implement/fix?
All tests are converted to be compatible with the Array API standard.

#### `jax.jit`
This PR is derived from #22777 where we already have a running Array API backend. To isolate changes to files, this PR already includes tests that check if the functions inside `scipy.spatial.transform` are compatible with `jax.jit`. In addition, three tests use `jax.jit` to accelerate testing, as the runtime without jitting increases significantly. Importantly, these changes do not influence the current behavior with numpy in any way.

#### Additional information
This PR does not enforce consistent formatting in the tests. Instead, it tries to match the existing code style to keep the amount of changes as small as possible and make reviewing easier. It would be great to have another PR after this one that makes clear which formatter should be used within `scipy.spatial.transform`. A general format style for `scipy` is not planned. See e.g. https://github.com/scipy/scipy/pull/22777#issuecomment-2827585033.

amacati committed:
TST: Add Array API tests to transform

amacati committed:
Merge branch 'scipy:main' into tst-transform

amacati committed:
Fix linting issues

amacati committed:
Unify use of pytest.raises

lucascolley commented:
> It would be great to have another PR after this one that makes clear which formatter should be used within scipy.spatial.transform. A general format style for scipy is not planned.

I'm happy to review a style-only PR 👍 just keep it separate from functional changes as you have planned

amacati commented:
> I'm happy to review a style-only PR 👍 just keep it separate from functional changes as you have planned

Will do once this one is merged. Thanks!

scottshambaugh commented on scipy/spatial/transform/tests/test_rigid_transform.py:
```diff
@@ -1,21 +1,37 @@
-def test_repr():
-    actual = repr(RigidTransform.identity())
+def rotation_to_xp(r: Rotation, xp):
+    return Rotation.from_quat(xp.asarray(r.as_quat()))
```
I'm unfamiliar with the array API, what are these two functions doing?

amacati replied:
`rotation_to_xp` is a helper that converts `Rotation` objects from `numpy` to `Rotation` objects with the correct Array API type. There are some functions in `Rotation` and `RigidTransform` that will always use numpy arrays, even in the new implementation. Examples are `Rotation.random` or `Rotation.identity` where it is impossible to decide which Array API framework should be used based on the input types, or where the Array API cannot provide a generic function (e.g. it does not specify how to create random numbers).

The conversion works like this:
`Rotation` in `numpy` -> `NDArray` with `as_quat()` -> `xp.Array` with `xp.asarray()` -> `Rotation` in `xp` with `R.from_quat()`

Long-term, we can change the input types to make it clear which framework to use, and then get rid of this helper.

lucascolley replied:
> it is impossible to decide which Array API framework should be used based on the input types

There is also the option of adding an `xp` kwarg rather than changing allowed input types, where it really does make sense to accept Python scalars rather than 0D arrays. See e.g. https://docs.scipy.org/doc/scipy/reference/generated/scipy.fft.fftfreq.html.

Either way, it can be done in a follow-up long-term, using some NumPy for now is fine.

amacati replied:
Yes, I was thinking along those lines. Some things like `random` will either need a specialized backend implementation or will have to use numpy though.

scottshambaugh commented on scipy/spatial/transform/tests/test_rigid_transform.py:
```diff
@@ -228,103 +302,126 @@ def test_from_exp_coords():
+    xp_assert_close(actual.as_matrix(), expected_matrix, atol=1e-12)
 
     # rotation and translation
+    # The tolerance is set to 1e-8, because xp_assert_close compares the absolute
```
Following up from the comments on the other PR, I think the tolerance change here is fine and doesn't need to be broken out to another PR. However I do believe that there is still a default rtol applied for floating point numbers, so this comment should change

scottshambaugh reviewed (approved):
I can confirm that all the tests are functionally unchanged and exercise the same behavior before and after this PR. Again, I'm not too familiar with the array API or the "why" behind these changes, but the changes themselves are no issue. The CI errors right now look like a setup issue rather than due to bad test values.

There's clearly an immense amount of effort here - well done! I'm good for a merge once the CI issues are worked through.

amacati committed:
Fix rtol comment

lucascolley commented:
A sample CI failures:

```
scipy/spatial/transform/tests/test_rotation.py:898: in test_as_euler_asymmetric_axes
rotation = Rotation.from_euler(seq, angles)
...
 _rotation.pyx:518: in scipy.spatial.transform._rotation._format_angles
     ???
...
E   TypeError: can't convert cuda:0 device type tensor to numpy. Use Tensor.cpu() to copy the tensor to host memory first.
```

amacati committed:
Disable xp backends

amacati commented:
> I guess we want to skip the whole modules for PyTorch and other backends for now?

Ah right, I thought that failed Array API tests are expected until we have merged the xp backend. They should be disabled now, let's see if the CI passes

lucascolley reviewed (commented):
thanks for the quick review @scottshambaugh !

I'll give @crusaderky some time to potentially check the JAX / lazy-specific sections, but otherwise we can probably merge this soon

crusaderky commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -897,139 +1068,199 @@ def test_as_euler_degenerate_compare_algorithms(seq_tuple, intrinsic):
-    assert_allclose(Rotation.identity(n).inv().magnitude(), 0)
+    r = Rotation.identity(n)
+    r = Rotation.from_quat(xp.asarray(r.as_quat()))
+    xp_assert_close(r.magnitude(), xp.asarray(0.0), check_shape=False)
+    xp_assert_close(r.inv().magnitude(), xp.asarray(0.0), check_shape=False)
```
Why can't we test the output shape?

amacati replied:
The previous assert just checked against 0, so it also didn't cover the shape. Using `check_shape=False` replicates that as closely as possible. I have changed it now to include the shape using `xp.zeros(n)`.

crusaderky commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -897,139 +1068,199 @@ def test_as_euler_degenerate_compare_algorithms(seq_tuple, intrinsic):
+    r = Rotation.from_quat(xp.eye(4))
     result = r.magnitude()
-    assert_array_almost_equal(result, [np.pi, np.pi, np.pi, 0])
+    xp_assert_close(result, xp.asarray([np.pi, np.pi, np.pi, 0]))
```
```suggestion
    xp_assert_close(result, xp.asarray([xp.pi, xp.pi, xp.pi, 0]))
```

crusaderky commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -897,139 +1068,199 @@ def test_as_euler_degenerate_compare_algorithms(seq_tuple, intrinsic):
+    r = Rotation.from_quat(xp.eye(4))
     result1 = r[0].magnitude()
-    assert_allclose(result1, np.pi)
+    xp_assert_close(result1, np.pi)
```
```suggestion
    xp_assert_close(result1, xp.pi)
```

crusaderky commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -1040,40 +1271,68 @@ def test_approx_equal_single_rotation():
+    axes = xp.concat((-xp.eye(3), xp.eye(3)))
     axes = np.concatenate((-np.eye(3), np.eye(3)))
-    thetas = np.linspace(0, np.pi / 2, 100)
+    thetas = xp.linspace(0, np.pi / 2, 100)
```
```suggestion
    thetas = xp.linspace(0, xp.pi / 2, 100)
```

crusaderky replied:
```suggestion
    thetas = xp.linspace(0, xp.pi / 2, 100)
```

crusaderky commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -1303,80 +1622,86 @@ def test_random_rotation_shape():
     rot, rssd, sens = Rotation.align_vectors(a, b, return_sensitivity=True)
-    assert np.isclose(rssd, rssd_expected, atol=atol)
-    assert np.allclose(sens, sens_expected, atol=atol)
+    assert xp.all(xpx.isclose(rssd, rssd_expected, atol=atol, xp=xp))
```
```suggestion
    xp_assert_close(rssd, rssd_expected, atol=atol)
```

crusaderky commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -1303,80 +1622,86 @@ def test_random_rotation_shape():
-    scale = 2
+    a = xp.asarray(Rotation.random(n, rng=0).apply([1, 0, 0]))
+    b = xp.asarray(Rotation.random(n, rng=1).apply([1, 0, 0]))
+    scale = xp.asarray(2.0)
```
```suggestion
    scale = 2
```

crusaderky commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -1303,80 +1622,86 @@ def test_random_rotation_shape():
-    assert_allclose(np.sqrt(scale) * rssd1, rssd2, atol=1e-6)
-    assert_allclose(cov1, cov2)
+    xp_assert_close(est1.as_matrix(), est2.as_matrix())
+    xp_assert_close(xp.sqrt(scale) * rssd1, rssd2, atol=1e-6)
```
```suggestion
    xp_assert_close(math.sqrt(scale) * rssd1, rssd2, atol=1e-6)
```

crusaderky commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -1388,99 +1713,137 @@ def test_align_vectors_noise():
+    m_expected = xp.asarray([[0.0, 0, 1],
+                             [1, 0, 0],
+                             [0, 1, 0]])
+    R, rssd = Rotation.align_vectors(a, b, weights=xp.asarray([np.inf, 1]))
```
```suggestion
    R, rssd = Rotation.align_vectors(a, b, weights=xp.asarray([xp.inf, 1]))
```

crusaderky commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -1388,99 +1713,137 @@ def test_align_vectors_noise():
+    R, rssd = Rotation.align_vectors(a, b, weights=xp.asarray([np.inf, 1]))
+    xp_assert_close(R.as_matrix(), m_expected, atol=atol)
+    xp_assert_close(R.apply(b), a, atol=atol)  # Pri and sec align exactly
+    assert xpx.isclose(rssd, xp.asarray(0.0), atol=atol, xp=xp)
```
```suggestion
    assert xpx.isclose(rssd, 0, atol=atol, xp=xp)
```

crusaderky commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -1388,99 +1713,137 @@ def test_align_vectors_noise():
-    assert np.isclose(rssd, rssd_expected, atol=atol)
+    b = xp.asarray([[1, 2, 3], [-2, 3, -1]])
+    a = xp.asarray([[-1.0, 3, 2], [1, -1, 2]])
+    rssd_expected = xp.asarray(1.3101595297515016)
+    R, rssd = Rotation.align_vectors(a, b, weights=xp.asarray([np.inf, 1]))
```
```suggestion
    rssd_expected = 1.3101595297515016
    R, rssd = Rotation.align_vectors(a, b, weights=xp.asarray([xp.inf, 1]))
```

crusaderky commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -1489,60 +1852,80 @@ def test_align_vectors_near_inf():
+    as_to_test = np.array([[[1.0, 0, 0], [0, 1, 0]],
                            [[0, 1, 0], [1, 0, 0]],
                            [[0, 0, 1], [0, 1, 0]]])
+
+    # We jit these tests for JAX because the non-jitted version significantly slows down
+    # the test
+    align_vectors = Rotation.align_vectors
+    if is_jax(xp):
+        import jax  # Must be installed because is_jax succeeds
+
+        align_vectors = jax.jit(Rotation.align_vectors)
+
```
```suggestion
```
ditto

crusaderky commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -1551,247 +1934,340 @@ def test_align_vectors_antiparallel():
     atol = 1e-12
     mats_a = Rotation.random(100, rng=0).as_matrix()
     mats_b = Rotation.random(100, rng=1).as_matrix()
+
+    # We jit these tests for JAX because the non-jitted version significantly slows down
+    # the test
+    align_vectors = Rotation.align_vectors
+    if is_jax(xp):
+        import jax  # Must be installed because is_jax succeeds
+
+        align_vectors = jax.jit(Rotation.align_vectors)
+
```
```suggestion
```

crusaderky commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -1551,247 +1934,340 @@ def test_align_vectors_antiparallel():
-    q = np.array([[0, 1, 0, 1], [0, 0, 1, 1]]) / np.sqrt(2)
+@pytest.mark.array_api_backends
+def test_repr_rotation_sequence(xp):
+    q = xp.asarray([[0.0, 1, 0, 1], [0, 0, 1, 1]]) / np.sqrt(2)
```
```suggestion
    q = xp.asarray([[0.0, 1, 0, 1], [0, 0, 1, 1]]) / math.sqrt(2)
```

crusaderky commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -1821,20 +2297,23 @@ def test_rotation_within_numpy_array():
 
 
 def test_deepcopy():
+    # TODO: Can we support this for all Array API frameworks?
```
```suggestion
    # Note: Array API makes no provision for arrays to support the `__copy__`
    # protocol, so it's OK to skip this test for the backends that don't
```

crusaderky commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -1976,23 +2502,39 @@ def test_as_davenport_degenerate():
-            with pytest.warns(UserWarning, match="Gimbal lock"):
-                angles_dav = rot.as_davenport(ax, order)
+            rot = Rotation.from_davenport(xp.asarray(ax), order, angles)
+            if is_lazy_array(rot.as_quat()):
+                angles_dav = rot.as_davenport(xp.asarray(ax), order)
+            else:  # We can only warn in non-lazy frameworks.
+                with pytest.warns(UserWarning, match="Gimbal lock"):
+                    angles_dav = rot.as_davenport(xp.asarray(ax), order)
```
```suggestion
            with eager_warns(rot_as_quat(), UserWarning, match="Gimbal lock"):
                angles_dav = rot.as_davenport(xp.asarray(ax), order)
```

crusaderky reviewed (changes requested):
- `test_*_jax_compile` are problematic for the reasons listed here: https://github.com/scipy/scipy/pull/22777#discussion_r2057765298
  What I would like to see:
  
  - For all these tests to be removed
  - Have all methods decorated with `@xp_capabilities`
  - At the top of each test module, add `lazy_xp_modules = [Rotation, RigidTransform]`
  - Have all tests decorated with `@make_skip_xp_backends(Rotation.<method>, ...)`

  _However_, this pattern

  - requires https://github.com/data-apis/array-api-extra/pull/284 to work (it should be merged any day now);
  - is untested with instance methods and class methods (need to add unit tests to array-api-extra);
  - I expect won't work with the current Cython classes.
  
  In the interest of pushing this PR through, I ask for now to just delete all `test_*_jax_compile` and leave JAX-specific treatment to a follow-up.

- Lazy backends don't have any way of eagerly validating inputs, as per the pattern established by `scipy.cluster.is_valid_linkage`.
  In the interest of pushing this PR through, I'd be content to have a follow-up issue to track the problem. 

- These methods, and probably a few more, have lost all test cases for when the input is a scalar or array-like:

  - from_euler
  - from_matrix
  - from_quat
  - from_translation
  - align_vectors
  - `__getitem__`

  Please add specific test cases for them (unless they're already there and I missed them).

- A bunch of minor observations as noted inline.

lucascolley commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -18,732 +30,869 @@ def basis_vec(axis):
-    x = np.array([
-        [3, 0, 0, 4],
+    x = xp.asarray([
+        [3.0, 0, 0, 4],
```
Typically these sorts of changes are fine if an extra test is added per function for dtype preservation

amacati commented on scipy/spatial/transform/tests/test_rigid_transform.py:
```diff
@@ -29,176 +48,234 @@ def test_repr():
-        RigidTransform.from_matrix(invalid)
+    invalid = xp.eye(4)
+    invalid = xpx.at(invalid)[..., 3, 3].set(2)  # Invalid last row
+    if is_lazy_array(invalid):
+        tf = RigidTransform.from_matrix(invalid)
+        assert xp.all(xp.isnan(tf.as_matrix()))
+    else:
+        with pytest.raises(ValueError):
+            RigidTransform.from_matrix(invalid)
```
Are you suggesting to add something like a `RigidTransform.is_valid_matrix`? That should be fairly easy to add for all options. However, where would you put this addition? The lazy distinctions (and `..jax_compile`) in the tests were only made to have all test changes in one commit. This would be a code change.

I could also remove the special lazy handling for now, since the tests are just supposed to work for numpy (as of now). If we add new methods to `RigidTransform` and `Rotation`, the PRs start to get mixed up again.

crusaderky replied:
As said elsewhere, I'm happy to have such a change discussed and implemented separately.

> I could also remove the special lazy handling for now

The point of this PR is to keep the size of the main PR down.
The moment you add lazy backends you'll need these code branches anyway.

To clarify: these changes to the already existing tests are OK, but they highlight a missing feature (eager validation) that impacts lazy backends only.

amacati commented on scipy/spatial/transform/tests/test_rigid_transform.py:
```diff
@@ -1,21 +1,40 @@
+    return RigidTransform.from_matrix(xp.asarray(r.as_matrix()))
+
+
+@pytest.mark.array_api_backends
```
Oh, good to know, thanks! Will remove them everywhere. Sometimes the pytest magic of inserting fixtures is a bit too convoluted for my taste..

amacati commented on scipy/spatial/transform/tests/test_rigid_transform.py:
```diff
@@ -228,103 +305,126 @@ def test_from_exp_coords():
+    xp_assert_close(actual.as_matrix(), expected_matrix, atol=1e-12)
 
     # rotation and translation
+    # The tolerance is set to 1e-8 because xp_assert_close deviates from 
+    # np.testing.assert_allclose in that it does not automatically default to 1e-7 for
+    # floating point inputs.
+    # See https://numpy.org/doc/2.2/reference/generated/numpy.testing.assert_allclose.html
```
At the moment, the line reads https://github.com/scipy/scipy/blob/c41054dfd5f3779bc350d9896191df10efe6bd91/scipy/spatial/transform/tests/test_rigid_transform.py#L336
which explicitly sets `atol`. If we do set `rtol` here, I assume we'd have to check for fp32 / fp64 manually, which also feels off. The comment was mainly meant to remind me where the difference in behaviour between xp and np stems from, because I wasn't aware of it before. I wasn't sure which combinations of `rtol` and `atol` would make fp32 computations pass, so I went for the easiest solution by reducing `atol`. Maybe this should be left at `1e-12` for now and revisited once the Array API backend and fp32 become relevant?

amacati commented on scipy/spatial/transform/tests/test_rigid_transform.py:
```diff
@@ -333,520 +433,599 @@ def test_from_dual_quat():
-    assert pytest.approx(np.linalg.norm(unnormalized_dual_quat[:4])) == 1.3
-    assert pytest.approx(np.dot(unnormalized_dual_quat[:4],
-                                unnormalized_dual_quat[4:]), abs=8) == 0.0
+    xp_assert_close(xp_vector_norm(unnormalized_dual_quat[:4]), xp.asarray(1.3),
+                    atol=1e-12, check_0d=False)
+    xp_assert_close(xp.vecdot(unnormalized_dual_quat[:4], unnormalized_dual_quat[4:]),
+                    xp.asarray(0.0), atol=1e-8, check_0d=False)
```
The result is not exactly zero, so the tolerances are needed, but I'll get rid of `check_0d`. Just out of curiosity: Why does `check_0d` exist if we don't use it in the tests?

lucascolley replied:
FWIW I wouldn't mind `check_0d` here. Probably cleaner without. But there is also the argument that `[()]` is undefined behaviour on general arrays (https://github.com/data-apis/array-api/issues/833).

> Why does check_0d exist if we don't use it in the tests?

Maybe it doesn't need to exist anymore, if it is really unused.

The history of 0d checking with these assertions is long and complicated—https://github.com/scipy/scipy/issues/21044 for the interested reader.

crusaderky replied:
As a final user, I would treat it as a regression if a function that started to return a np.ndarray all of a sudden while until the previous version it returned a np.float64. Namely, everything that relies on the returned value to be a subclass of float would break; everything that expects it to be hashable would follow the same fate.

I agree that we should reconsider the usefulness of check_0d. To me it makes sense only in a case where you can't predict if you're going to receive an array or a generic, but... you can always do that even if you don't know the shape of your `expect`'ed array by just applying `[()]` to it.

amacati commented on scipy/spatial/transform/tests/test_rigid_transform.py:
```diff
@@ -333,520 +433,599 @@ def test_from_dual_quat():
-    random_dual_quats[:, :4] = random_dual_quats[:, :4].round(2)
-    assert not np.any(np.isclose(
-        np.linalg.norm(random_dual_quats[:, :4], axis=1), 1.0, atol=0.0001))
+    non_normal_dual_quats = random_dual_quats[:, :4] + 0.01
+    random_dual_quats = xpx.at(random_dual_quats)[:, :4].set(non_normal_dual_quats)
```
Yes. The rounding before was to make sure there is a deviation from the normalized (i.e. norm 1) quaternions. Since `round()` was not available, the new version adds a constant to the quaternion to achieve the same effect. The assert afterwards double-checks that we really have non-normalized quaternions.

amacati commented on scipy/spatial/transform/tests/test_rigid_transform.py:
```diff
@@ -860,86 +1039,105 @@ def test_indexing_validation():
 
 
-def test_copy_flag():
+@pytest.mark.skip_xp_backends(eager_only=True,
+                              reason="lazy backends do not support memory sharing")
```
Thanks for the correction. I've only worked with jax as lazy framework before and was assuming that other lazy frameworks have the same restriction on mutability.

amacati commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -18,732 +30,869 @@ def basis_vec(axis):
-    mat = np.array([
+@pytest.mark.array_api_backends
+def test_from_matrix_calculation(xp):
+    atol = 1e-8
```
No, but same as the other one I'm not sure which settings are preferred here. `atol` was just used before to control the precision, so I stuck with it. This should only be required for fp32

amacati commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -1489,60 +1852,80 @@ def test_align_vectors_near_inf():
     for i in range(6):
         mats.append(Rotation.random(n, rng=10 + i).as_matrix())
 
+    # We jit these tests for JAX because the non-jitted version significantly slows down
+    # the test
+    align_vectors = Rotation.align_vectors
+    if is_jax(xp):
+        import jax  # Must be installed because is_jax succeeds
+
+        align_vectors = jax.jit(Rotation.align_vectors)
+
```
I believe these are not yet ready, right? Will remove the jit anyways

amacati commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -1551,247 +1934,340 @@ def test_align_vectors_antiparallel():
+    q = xp.asarray([0, 0, 0, 1])
     actual = repr(Rotation.from_quat(q))
-    expected = """\
+    # TODO: Decide how we want to print that the inner array may be a different type
+    # (e.g. a torch.Tensor)
```
What is your opinion on this? Ideally the xp framework would be visible in the print

amacati commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -1886,61 +2381,90 @@ def test_len_and_bool():
-    rot = Rotation.from_rotvec([np.array(ez) * np.pi/4,
-                                np.array(ey) * np.pi/6])
+    axes = xp.stack([ez, ey], axis=0)
+    rot = Rotation.from_rotvec(axes * xp.asarray([[np.pi/4], [np.pi/6]]))
```
Done

amacati committed:
Address review comments

amacati commented:
I have addressed most of the review comments by removing the jax.jit tests, adding tests for array_like inputs and scalars where applicable, and fixing the inline comments from @crusaderky 

The main remaining item is the change of tolerances for some tests. I'm not sure what your preference or policy is there

amacati committed:
Fix linting. Fix Array-ness errors

amacati committed:
Fix linter

crusaderky commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -1236,47 +1417,62 @@ def test_getitem():
 
-def test_setitem_single():
-    r = Rotation.identity()
+def test_getitem_array_like(xp):
+    mat = np.array([[[0.0, -1, 0],
+                     [1, 0, 0],
+                     [0, 0, 1]],
+
+                    [[1, 0, 0],
+                     [0, 0, -1],
+                     [0, 1, 0]]])
+    mat = xp.asarray(mat)
+    r = Rotation.from_matrix(mat)
+    xp_assert_close(r[[0]].as_matrix(), mat[[0]], atol=1e-15)
+    xp_assert_close(r[[0, 1]].as_matrix(), mat[[0, 1]], atol=1e-15)
```
I'd rather leave this behaviour untested for xp != np. Intuitively, this should be aligned to the Array API `__getitem__` where array-likes are not supported (read: their behaviour is undefined).
```suggestion
def test_getitem_array_like():
    mat = np.array([[[0.0, -1, 0],
                     [1, 0, 0],
                     [0, 0, 1]],
                    [[1, 0, 0],
                     [0, 0, -1],
                     [0, 1, 0]]])
    r = Rotation.from_matrix(mat)
    xp_assert_close(r[[0]].as_matrix(), mat[[0]], atol=1e-15)
    xp_assert_close(r[[0, 1]].as_matrix(), mat[[0, 1]], atol=1e-15)
```

crusaderky commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -1821,20 +2112,25 @@ def test_rotation_within_numpy_array():
+def test_pickling(xp):
+    # Note: Array API makes no provision for arrays to be pickleable, so
+    # it's OK to skip this test for the backends that don't support it
+    r = Rotation.from_quat(xp.asarray([0, 0, math.sin(xp.pi/4), math.cos(xp.pi/4)]))
```
```suggestion
    r = Rotation.from_quat(xp.asarray([0, 0, math.sin(np.pi/4), math.cos(np.pi/4)]))
```

crusaderky commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -1821,20 +2112,25 @@ def test_rotation_within_numpy_array():
+    xp_assert_close(r.as_matrix(), unpickled.as_matrix(), atol=1e-15)
 
 
 def test_deepcopy():
```
```suggestion
def test_deepcopy(xp):
```

crusaderky commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -1821,20 +2112,25 @@ def test_rotation_within_numpy_array():
-    r = Rotation.from_quat([0, 0, np.sin(np.pi/4), np.cos(np.pi/4)])
+    # Note: Array API makes no provision for arrays to support the `__copy__`
+    # protocol, so it's OK to skip this test for the backends that don't
+    r = Rotation.from_quat([0, 0, math.sin(np.pi/4), math.cos(np.pi/4)])
```
```suggestion
    r = Rotation.from_quat(xp.asarray([0, 0, math.sin(np.pi/4), math.cos(np.pi/4)]))
```

crusaderky commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -1886,180 +2185,215 @@ def test_len_and_bool():
-    angles_middle = [rnd.choice([0, np.pi]) for i in range(n)]
-    angles[:, 2] = rnd.uniform(low=-np.pi, high=np.pi, size=(n,))
-    lambdas = rnd.uniform(low=0, high=np.pi, size=(5,))
+    angles[:, 0] = rnd.uniform(low=-xp.pi, high=xp.pi, size=(n,))
+    angles_middle = [rnd.choice([0, xp.pi]) for i in range(n)]
+    angles[:, 2] = rnd.uniform(low=-xp.pi, high=xp.pi, size=(n,))
+    lambdas = rnd.uniform(low=0, high=xp.pi, size=(5,))
```
```suggestion
    angles[:, 0] = rnd.uniform(low=-np.pi, high=np.pi, size=(n,))
    angles_middle = [rnd.choice([0, np.pi]) for i in range(n)]
    angles[:, 2] = rnd.uniform(low=-np.pi, high=np.pi, size=(n,))
    lambdas = rnd.uniform(low=0, high=np.pi, size=(5,))
```

crusaderky commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -1886,180 +2185,215 @@ def test_len_and_bool():
-    angles[:, 0] = rnd.uniform(low=-np.pi, high=np.pi, size=(n,))
-    angles[:, 1] = rnd.uniform(low=0, high=np.pi, size=(n,))
-    angles[:, 2] = rnd.uniform(low=-np.pi, high=np.pi, size=(n,))
+    angles[:, 0] = rnd.uniform(low=-xp.pi, high=xp.pi, size=(n,))
+    angles[:, 1] = rnd.uniform(low=0, high=xp.pi, size=(n,))
+    angles[:, 2] = rnd.uniform(low=-xp.pi, high=xp.pi, size=(n,))
```
```suggestion
    angles[:, 0] = rnd.uniform(low=-np.pi, high=np.pi, size=(n,))
    angles[:, 1] = rnd.uniform(low=0, high=np.pi, size=(n,))
    angles[:, 2] = rnd.uniform(low=-np.pi, high=np.pi, size=(n,))
```

crusaderky commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -1886,180 +2185,215 @@ def test_len_and_bool():
 
     # asymmetric sequences
-    angles[:, 1] -= np.pi / 2
+    angles[:, 1] -= xp.pi / 2
```
```suggestion
    angles[:, 1] -= np.pi / 2
```

crusaderky commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -1886,180 +2185,215 @@ def test_len_and_bool():
-    angles[:, 0] = rnd.uniform(low=-np.pi, high=np.pi, size=(n,))
-    angles[:, 1] = rnd.uniform(low=0, high=np.pi, size=(n,))
-    angles[:, 2] = rnd.uniform(low=-np.pi, high=np.pi, size=(n,))
+    angles[:, 0] = rnd.uniform(low=-xp.pi, high=xp.pi, size=(n,))
+    angles[:, 1] = rnd.uniform(low=0, high=xp.pi, size=(n,))
+    angles[:, 2] = rnd.uniform(low=-xp.pi, high=xp.pi, size=(n,))
```
```suggestion
    angles[:, 0] = rnd.uniform(low=-np.pi, high=np.pi, size=(n,))
    angles[:, 1] = rnd.uniform(low=0, high=np.pi, size=(n,))
    angles[:, 2] = rnd.uniform(low=-np.pi, high=np.pi, size=(n,))
```

crusaderky commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -1886,180 +2185,215 @@ def test_len_and_bool():
 
     # asymmetric sequences
-    angles[:, 1] -= np.pi / 2
+    angles[:, 1] -= xp.pi / 2
```
```suggestion
    angles[:, 1] -= np.pi / 2
```

crusaderky reviewed (changes requested):
I don't have a very strong opinion re. tolerances.
On one hand, yes it would be nice if they were in a separate, specific PR; on the other hand I don't want to add unnecessary attrition.

So, I defer the decision to you.

A bunch of nits below. I'll be away until next Thursday; please don't wait for my formal thumbs up to merge.

amacati commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -18,732 +32,847 @@ def basis_vec(axis):
-    angles[:, 0] = rnd.uniform(low=-np.pi, high=np.pi, size=(n,))
-    angles[:, 1] = rnd.uniform(low=-np.pi / 2, high=np.pi / 2, size=(n,))
-    angles[:, 2] = rnd.uniform(low=-np.pi, high=np.pi, size=(n,))
+    angles[:, 0] = rnd.uniform(low=-xp.pi, high=xp.pi, size=(n,))
+    angles[:, 1] = rnd.uniform(low=-xp.pi / 2, high=xp.pi / 2, size=(n,))
+    angles[:, 2] = rnd.uniform(low=-xp.pi, high=xp.pi, size=(n,))
```
No worries, it makes sense to be strict. This was changed by a bulk search and rename. Good catch

purepani replied:
> This is just defensive in case one day some backend defines pi as something else.

Would be great for an april fools backend :smile: : https://en.wikipedia.org/wiki/Indiana_pi_bill

crusaderky replied:
I was actually implying the possibility of a `float128`, or a ML backend defining it as `float32` or even `float16` or `bfloat16`, but `3.14` will do too I suppose 😆

purepani replied:
Not even that! It was implied to be 3.2 in that bill!

amacati commented on scipy/spatial/transform/tests/test_rotation.py:
```diff
@@ -1224,7 +1388,24 @@ def test_apply_shapes():
 
-def test_getitem():
+
+def test_apply_array_like(xp):
+    rng = np.random.default_rng(123)
+    # Single vector
+    r = rotation_to_xp(Rotation.random(rng=rng), xp)
+    t = rng.uniform(-100, 100, size=(3,)).tolist()
+    v = r.apply(t)
+    v_expected = r.apply(xp.asarray(t))
+    xp_assert_close(v, v_expected, atol=1e-12)
+
+    # Multiple vectors
+    t = rng.uniform(-100, 100, size=(2, 3)).tolist()
+    v = r.apply(t)
+    v_expected = r.apply(xp.asarray(t))
+    xp_assert_close(v, v_expected, atol=1e-12)
```
This change was intentional. The thought here was that if we already have an xp rotation, we may want to be able to pass in lists to those as well. But it also makes sense to only test this for numpy. After all, passing lists when dealing with xp other than numpy is most likely unintentional

amacati committed:
Fix remaining precision issues. Use xp_default_dtype

amacati commented:
Okay, I think this is ready to be merged if the CI passes. Thanks a lot @lucascolley, @scottshambaugh and @crusaderky for the detailed reviews. That was a lot of code to work through. Now that the tests are in place we can hopefully start to merge the Array API backend soon.

amacati commented:
By the way, I see that `xp_default_dtype` returns `float64` for all backends but torch. How is this consistent with `jax`'s default precision of `float32`?

lucascolley commented:
https://docs.jax.dev/en/latest/default_dtypes.html#the-x64-flag-enabling-64-bit-values, https://github.com/scipy/scipy/blob/main/scipy/conftest.py#L193. It seems that JAX effectively offers 2 modes: either you cannot use 64-bit dtypes at all, or they are the default.

lucascolley reviewed (approved):
This has been thoroughly reviewed at this point and is passing CI—thanks a lot @amacati ! Excited for the follow-ups!

amacati commented:
> https://github.com/scipy/scipy/blob/main/scipy/conftest.py#L193. It seems that JAX effectively offers 2 modes: either you cannot use 64-bit dtypes at all, or they are the default.

I wasn't aware `scipy` sets the f64 variable for testing. This might be of interest for the coming PRs. Since `jax` is usually running on f32, it would make sense to also test that. In addition, the Array API backend draft is currently designed to preserve f32 for **all** backends but `numpy` (because that would be a change in behaviour from the current implementation). I.e., a `Rotation` created with a `torch.float32` tensor will preserve the `dtype` and calculate in single precision.

This design makes most sense to me, but we'll probably have to discuss this decision once the backend PR is ready.

lucascolley commented:
1. I think there is definitely room to make 'no f64 support assumed' work in some parts of SciPy
2. Doing that for the whole of SciPy seems like a monumental task. I think some parts just won't work 
3. Differing dtype behaviour for NumPy and alternative backends may be justified, but we will have to consider this again when we come to it.

On (2), I think we probably want to wait for some movement on https://github.com/data-apis/array-api-compat/issues/162. It would be interesting to see the quantity if failures if we implemented a 32-bit dtype which 'pretends' to be `float64` (with some sort of shim to expect 32-bit precision results in tests).

crusaderky commented:
> 1. I think there is definitely room to make 'no f64 support assumed' work in some parts of SciPy
> 2. Doing that for the whole of SciPy seems like a monumental task. I think some parts just won't work

Worth pointing out that without that flag JAX is falling foul of the Array API specs. That's very different from PyTorch's default type, which impacts exactly just that and which is covered by the Standard. 

> 3. Differing dtype behaviour for NumPy and alternative backends may be justified, but we will have to consider this again when we come to it.

The default dtype being float32 instead of float64 is covered. The rest shouldn't IMHO.

lucascolley commented:
> I assume that's because type promotions etc do not work the way one expects them to in fp32 mode?

It's more simple than that—the standard requires that e.g. `float64` is supported (https://data-apis.org/array-api/draft/API_specification/data_types.html#api-specification-data-types--page-root). I think it was briefly discussed at some point to figure out how to relax that requirement, especially with MLX being 32-bit only. But I don't think much work has happened in that direction.
